circuits/circuits/claiming/distribute.circom

```diff
@@ -0,0 +1,229 @@
+            claims <-- [sorted_claims_hack[i-1], sorted_claims_hack[i]] // claims is advice that is constrained to correspond to the leaf in the component
+            // TODO: pass claims in like this
+            // claims <-- [claims[reverse_permutation_Indices[i-1]], claims[reverse_permutation_Indices[i]]]
+        );
+    }
+}
+
+// ClaimsInOrder verifies that a list of leaves strictly ascends as follows, where H is a hash function:
+// H([0-10]), H([12-15]) ... H([1001-1005]), 0, 0, ..., 0
+//
+// This requires two checks on each pair of elements:
+//      1/ Zeroes must be followed by zeroes
+//      2/ Non-zero values must be ordered
+// 
+// CAUTION: when we move to Nova, we will probably halt before checking every possible leaf, so we must make sure that the unchecked leaves are also zero.
+// This can be done by checking that for some zero leaf, all pathElements where the pathIndex is 0 (i.e., our element is on the left) have the appropriate value for
+// their level. We can cheaply precalculate the proper value as p_k = h(p_{k+1}, p_{k+1}) where p_n = 0 for a depth n tree by caching each p_k.
+template ClaimsInOrder() {
+    signal input leaves[2];
+    // claims is unconstrained advice, since we conditionally check that the claim hashes to the leaf in the component
+    // It needs to be passed the correct preimages for non-zero nodes or witness generation will fail
+    signal input claims[2][3]; 
+
+    // --- Step 1 ----
+    // If we have a zeroed value, make sure it's followed by a zero
+    // i.e., if the previous value is zero, the current value can't be non-zero: NAND(prev_is_zero, curr_is_nonzero)
+    signal curr_is_zero <== IsZero()(leaves[1]);
+    signal curr_is_nonzero <== 1 - curr_is_zero;
+    signal prev_is_zero <== IsZero()(leaves[0]);
+
+    signal following_valid <== NAND()(
+        a <== prev_is_zero,
+        b <== curr_is_nonzero
+    );
+    following_valid === 1;
+
+    // ---- Step 2 ----
+    // Ensure claims are in order
+    // Note, since this component is used on the filtered tree, each claim should already have been validated, so claims[i][1] <= claims[i][2] for all i
+    // Set passing_claims to values that pass the ordering check
+    signal passing_claims[2][3];
+    var trivially_passing[2][3] = [[0,0,0],[0,1,0]]; // Upper end of lower range is 0, lower end of upper range is 1
+    for (var i = 0; i < 2; i++) {
+        for (var j = 0; j < 3; j++) {
+            passing_claims[i][j] <-- curr_is_nonzero ? claims[i][j] : trivially_passing[i][j];
+        }
+    }
+    signal calculated_lt <== LessThan(128)(
+        [passing_claims[0][2],
+        passing_claims[1][1]]
+    );
+    calculated_lt === 1;
+    
+    // Ensure that the claims hash to the provided leaves for all non-zero leaves
+    signal calculated_leaf_a <== Poseidon(3)(claims[0]);
+    signal calculated_leaf_b <== Poseidon(3)(claims[1]);
+    curr_is_nonzero * (calculated_leaf_a - leaves[0]) === 0;
```

Comment:
shouldn't this be `(1-prev_is_zero) * (calculated_leaf_a - leaves[0]) === 0;`?

Reply:
I think that works too, but I don't see a bug here.
Like, we only need to enforce the relationship between the two if the current is value is non zero. If the previous is non zero, and the current is zero, we still don't have to enforce any relationship between them.